File: .opencode/skills/rlm/parallel-processor.py
"""
Parallel Wave Processor for RLM
Based on: MIT Recursive Language Models paper (arXiv:2512.24601)

Purpose: 3-4x speedup for large context processing
"""

import asyncio
import os
import threading
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime
from typing import Any, Callable, Dict, List

# Constants
MAX_CONTENT_LENGTH = 5000

# Singleton
_processor_instance = None
_instance_lock = threading.Lock()


def get_parallel_processor(max_concurrent: int = 5, min_confidence: float = 0.8):
    """Get singleton processor instance"""
    global _processor_instance
    if _processor_instance is None:
        with _instance_lock:
            if _processor_instance is None:
                _processor_instance = ParallelWaveProcessor(
                    max_concurrent=max_concurrent,
                    min_confidence=min_confidence,
                )
    return _processor_instance


class ParallelWaveProcessor:
    """Process RLM chunks in parallel waves"""

    def __init__(
        self,
        max_concurrent: int = 5,
        min_confidence: float = 0.8,
        min_high_confidence_chunks: int = 3,
    ):
        self.max_concurrent = max_concurrent
        self.min_confidence = min_confidence
        self.min_high_confidence_chunks = min_high_confidence_chunks

    # -------------------------------------------------------------------------
    # Core processing (single implementation for sync/async)
    # -------------------------------------------------------------------------

    async def process_chunk(
        self,
        chunk_path: str,
        query: str,
        callback: Callable,
    ) -> Dict[str, Any]:
        """Process single chunk - works with sync or async callbacks"""
        chunk_id = os.path.basename(chunk_path).replace(".txt", "").replace(".json", "")

        try:
            with open(chunk_path, encoding="utf-8") as f:
                content = f.read()[:MAX_CONTENT_LENGTH]

            # Handle both sync and async callbacks
            if asyncio.iscoroutinefunction(callback):
                result = await callback(
                    {
                        "chunk_file": chunk_path,
                        "chunk_content": content,
                        "query": query,
                        "chunk_size": len(content),
                    }
                )
            else:
                result = callback(
                    {
                        "chunk_file": chunk_path,
                        "chunk_content": content,
                        "query": query,
                        "chunk_size": len(content),
                    }
                )

            return {
                "chunk_id": chunk_id,
                "success": True,
                "result": result,
                "timestamp": datetime.now().isoformat(),
                "tokens_processed": len(content.split()),
            }
        except Exception as e:
            return {
                "chunk_id": chunk_id,
                "success": False,
                "error": str(e),
                "timestamp": datetime.now().isoformat(),
                "tokens_processed": 0,
            }

    # -------------------------------------------------------------------------
    # Wave processing (async with sync wrapper)
    # -------------------------------------------------------------------------

    async def process_wave(
        self,
        chunk_paths: List[str],
        query: str,
        callback: Callable,
    ) -> List[Dict[str, Any]]:
        """Process wave of chunks in parallel"""
        paths = chunk_paths[: self.max_concurrent]

        if asyncio.iscoroutinefunction(callback):
            tasks = [self.process_chunk(p, query, callback) for p in paths]
            results = await asyncio.gather(*tasks, return_exceptions=True)
        else:
            # Sync path - use thread pool
            with ThreadPoolExecutor(max_workers=self.max_concurrent) as executor:
                futures = {
                    executor.submit(self._sync_wrapper, p, query, callback): p
                    for p in paths
                }
                results = [f.result() for f in as_completed(futures)]

        # Normalize exceptions
        normalized = []
        for r in results:
            if isinstance(r, Exception):
                normalized.append(
                    {
                        "chunk_id": "unknown",
                        "success": False,
                        "error": str(r),
                        "timestamp": datetime.now().isoformat(),
                        "tokens_processed": 0,
                    }
                )
            else:
                normalized.append(r)
        return normalized

    def _sync_wrapper(
        self, chunk_path: str, query: str, callback: Callable
    ) -> Dict[str, Any]:
        """Sync wrapper to run async process_chunk"""
        return asyncio.run(self.process_chunk(chunk_path, query, callback))

    # -------------------------------------------------------------------------
    # Main entry point
    # -------------------------------------------------------------------------

    async def process_all(
        self,
        chunk_paths: List[str],
        query: str,
        callback: Callable,
    ) -> Dict[str, Any]:
        """Process all chunks in waves with early termination"""
        waves = [
            chunk_paths[i : i + self.max_concurrent]
            for i in range(0, len(chunk_paths), self.max_concurrent)
        ]

        all_results = []
        wave_idx = 0
        for wave_idx, wave in enumerate(waves, 1):
            wave_results = await self.process_wave(wave, query, callback)
            all_results.extend(wave_results)

            if self._has_confident_answer(wave_results):
                break

        return {
            "total_waves": len(waves),
            "processed_waves": wave_idx,
            "total_chunks": len(chunk_paths),
            "processed_chunks": len(all_results),
            "successful_chunks": sum(1 for r in all_results if r["success"]),
            "results": all_results,
            "tokens_processed": sum(r.get("tokens_processed", 0) for r in all_results),
            "early_termination": wave_idx < len(waves),
            "confidence_threshold": self.min_confidence,
        }

    def process_all_chunks(
        self, chunk_paths: List[str], query: str, callback: Callable
    ) -> Dict[str, Any]:
        """Sync entry point"""
        return asyncio.run(self.process_all(chunk_paths, query, callback))

    # -------------------------------------------------------------------------
    # Helpers
    # -------------------------------------------------------------------------

    def _has_confident_answer(self, results: List[Dict]) -> bool:
        if not results:
            return False
        high_conf = [
            r
            for r in results
            if r.get("success") and self._extract_confidence(r) >= self.min_confidence
        ]
        return len(high_conf) >= self.min_high_confidence_chunks

    def _extract_confidence(self, result: Dict) -> float:
        conf = result.get("result", {}).get("confidence", 0.0)
        if isinstance(conf, (int, float)):
            return float(conf)
        if isinstance(result.get("result"), dict):
            for key in ["confidence", "certainty", "score", "probability"]:
                if key in result["result"] and isinstance(
                    result["result"][key], (int, float)
                ):
                    return float(result["result"][key])
        return 0.0

    def get_statistics(self) -> Dict[str, Any]:
        return {
            "max_concurrent": self.max_concurrent,
            "min_confidence": self.min_confidence,
            "min_high_confidence_chunks": self.min_high_confidence_chunks,
        }


# CLI
if __name__ == "__main__":
    import argparse
    import shutil
    import time

    parser = argparse.ArgumentParser(description="Parallel Wave Processor")
    subparsers = parser.add_subparsers(dest="command")

    test_parser = subparsers.add_parser("test")
    test_parser.add_argument("--chunks", type=int, default=10)
    test_parser.add_argument("--max-concurrent", type=int, default=5)

    args = parser.parse_args()

    if args.command == "test":
        test_dir = ".opencode/rlm_state/test_chunks"
        os.makedirs(test_dir, exist_ok=True)

        for i in range(args.chunks):
            with open(os.path.join(test_dir, f"chunk_{i:03d}.txt"), "w") as f:
                f.write(f"Test chunk {i}\n" * 100)

        chunk_paths = sorted(
            [
                os.path.join(test_dir, f)
                for f in os.listdir(test_dir)
                if f.endswith(".txt")
            ]
        )

        def mock_callback(data):
            return {
                "confidence": 0.9 if "5" in data["chunk_id"] else 0.6,
                "answer": f"Answer from {data['chunk_id']}",
            }

        processor = ParallelWaveProcessor(max_concurrent=args.max_concurrent)
        start = time.time()
        results = processor.process_all_chunks(chunk_paths, "test query", mock_callback)

        print(
            f"Processed {results['processed_chunks']} chunks in {results['processed_waves']} waves"
        )
        print(f"Time: {time.time() - start:.2f}s")
        print(f"Early termination: {results['early_termination']}")

        shutil.rmtree(test_dir)

File: .opencode/skills/rlm/parallel-processor.ts
#!/usr/bin/env bun
/**
 * Parallel Wave Processor for RLM
 * Based on: MIT Recursive Language Models paper (arXiv:2512.24601)
 * 
 * Purpose: 3-4x speedup for large context processing
 * 
 * Converted from: parallel_processor.py
 */

import { existsSync, readFileSync, rmSync, mkdirSync, writeFileSync } from 'fs';
import { join, basename } from 'path';
import { colors, printHeader } from '../../cli/lib/colors';

const MAX_CONTENT_LENGTH = 5000;

interface ChunkResult {
  chunk_id: string;
  success: boolean;
  result?: unknown;
  error?: string;
  timestamp: string;
  tokens_processed: number;
}

interface ProcessResult {
  total_waves: number;
  processed_waves: number;
  total_chunks: number;
  processed_chunks: number;
  successful_chunks: number;
  results: ChunkResult[];
  tokens_processed: number;
  early_termination: boolean;
  confidence_threshold: number;
}

interface ProcessStats {
  max_concurrent: number;
  min_confidence: number;
  min_high_confidence_chunks: number;
}

type CallbackFunction = (data: {
  chunk_file: string;
  chunk_content: string;
  query: string;
  chunk_size: number;
}) => unknown | Promise<unknown>;

class ParallelWaveProcessor {
  private maxConcurrent: number;
  private minConfidence: number;
  private minHighConfidenceChunks: number;

  constructor(
    maxConcurrent = 5,
    minConfidence = 0.8,
    minHighConfidenceChunks = 3
  ) {
    this.maxConcurrent = maxConcurrent;
    this.minConfidence = minConfidence;
    this.minHighConfidenceChunks = minHighConfidenceChunks;
  }

  async processChunk(
    chunkPath: string,
    query: string,
    callback: CallbackFunction
  ): Promise<ChunkResult> {
    const chunkId = basename(chunkPath).replace(/\.txt$/, '').replace(/\.json$/, '');

    try {
      const content = existsSync(chunkPath)
        ? readFileSync(chunkPath, 'utf-8').slice(0, MAX_CONTENT_LENGTH)
        : '';

      const result = await callback({
        chunk_file: chunkPath,
        chunk_content: content,
        query,
        chunk_size: content.length,
      });

      return {
        chunk_id: chunkId,
        success: true,
        result,
        timestamp: new Date().toISOString(),
        tokens_processed: content.split(/\s+/).length,
      };
    } catch (e) {
      return {
        chunk_id: chunkId,
        success: false,
        error: String(e),
        timestamp: new Date().toISOString(),
        tokens_processed: 0,
      };
    }
  }

  async processWave(
    chunkPaths: string[],
    query: string,
    callback: CallbackFunction
  ): Promise<ChunkResult[]> {
    const paths = chunkPaths.slice(0, this.maxConcurrent);
    const promises = paths.map(p => this.processChunk(p, query, callback));
    const results = await Promise.allSettled(promises);

    return results.map(r => {
      if (r.status === 'fulfilled') {
        return r.value;
      }
      return {
        chunk_id: 'unknown',
        success: false,
        error: r.reason?.message || String(r.reason),
        timestamp: new Date().toISOString(),
        tokens_processed: 0,
      };
    });
  }

  async processAll(
    chunkPaths: string[],
    query: string,
    callback: CallbackFunction
  ): Promise<ProcessResult> {
    const waves: string[][] = [];
    for (let i = 0; i < chunkPaths.length; i += this.maxConcurrent) {
      waves.push(chunkPaths.slice(i, i + this.maxConcurrent));
    }

    const allResults: ChunkResult[] = [];
    let waveIdx = 0;

    for (waveIdx = 0; waveIdx < waves.length; waveIdx++) {
      const waveResults = await this.processWave(waves[waveIdx], query, callback);
      allResults.push(...waveResults);

      if (this.hasConfidentAnswer(waveResults)) {
        waveIdx++; // Increment before break for accurate count
        break;
      }
    }

    return {
      total_waves: waves.length,
      processed_waves: waveIdx,
      total_chunks: chunkPaths.length,
      processed_chunks: allResults.length,
      successful_chunks: allResults.filter(r => r.success).length,
      results: allResults,
      tokens_processed: allResults.reduce((sum, r) => sum + r.tokens_processed, 0),
      early_termination: waveIdx < waves.length,
      confidence_threshold: this.minConfidence,
    };
  }

  private hasConfidentAnswer(results: ChunkResult[]): boolean {
    if (results.length === 0) return false;

    const highConf = results.filter(r => {
      if (!r.success) return false;
      return this.extractConfidence(r) >= this.minConfidence;
    });

    return highConf.length >= this.minHighConfidenceChunks;
  }

  private extractConfidence(result: ChunkResult): number {
    const res = result.result as Record<string, unknown> | undefined;
    if (!res) return 0;

    const conf = res.confidence;
    if (typeof conf === 'number') return conf;

    for (const key of ['confidence', 'certainty', 'score', 'probability']) {
      const val = res[key];
      if (typeof val === 'number') return val;
    }

    return 0;
  }

  getStatistics(): ProcessStats {
    return {
      max_concurrent: this.maxConcurrent,
      min_confidence: this.minConfidence,
      min_high_confidence_chunks: this.minHighConfidenceChunks,
    };
  }
}

let processorInstance: ParallelWaveProcessor | null = null;

function getParallelProcessor(
  maxConcurrent = 5,
  minConfidence = 0.8
): ParallelWaveProcessor {
  if (!processorInstance) {
    processorInstance = new ParallelWaveProcessor(maxConcurrent, minConfidence);
  }
  return processorInstance;
}

// CLI
function printUsage(): void {
  printHeader('Parallel Wave Processor');
  console.log();
  console.log(`${colors.yellow}Usage:${colors.reset}`);
  console.log(`  bun run parallel-processor.ts <command> [args]`);
  console.log();
  console.log(`${colors.yellow}Commands:${colors.reset}`);
  console.log(`  ${colors.green}test${colors.reset} [options]          Run test with mock chunks`);
  console.log();
  console.log(`${colors.yellow}Test Options:${colors.reset}`);
  console.log(`  --chunks <n>            Number of test chunks (default: 10)`);
  console.log(`  --max-concurrent <n>    Max concurrent processes (default: 5)`);
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);

  if (args.length === 0 || args[0] === '--help' || args[0] === '-h') {
    printUsage();
    process.exit(0);
  }

  const command = args[0];

  if (command === 'test') {
    let chunks = 10;
    let maxConcurrent = 5;

    for (let i = 1; i < args.length; i++) {
      if (args[i] === '--chunks' && args[i + 1]) {
        chunks = parseInt(args[++i], 10);
      } else if (args[i] === '--max-concurrent' && args[i + 1]) {
        maxConcurrent = parseInt(args[++i], 10);
      }
    }

    const testDir = '.opencode/rlm_state/test_chunks';
    mkdirSync(testDir, { recursive: true });

    // Create test chunks
    for (let i = 0; i < chunks; i++) {
      const chunkPath = join(testDir, `chunk_${String(i).padStart(3, '0')}.txt`);
      writeFileSync(chunkPath, `Test chunk ${i}\n`.repeat(100));
    }

    const chunkPaths: string[] = [];
    for (let i = 0; i < chunks; i++) {
      chunkPaths.push(join(testDir, `chunk_${String(i).padStart(3, '0')}.txt`));
    }

    const mockCallback: CallbackFunction = (data) => {
      return {
        confidence: data.chunk_id.includes('5') ? 0.9 : 0.6,
        answer: `Answer from ${data.chunk_id}`,
      };
    };

    const processor = new ParallelWaveProcessor(maxConcurrent);
    const start = Date.now();
    const results = await processor.processAll(chunkPaths, 'test query', mockCallback);

    console.log(`Processed ${results.processed_chunks} chunks in ${results.processed_waves} waves`);
    console.log(`Time: ${((Date.now() - start) / 1000).toFixed(2)}s`);
    console.log(`Early termination: ${results.early_termination}`);

    // Cleanup
    rmSync(testDir, { recursive: true, force: true });
  } else {
    console.error(`${colors.red}Unknown command: ${command}${colors.reset}`);
    printUsage();
    process.exit(1);
  }
}

export { ParallelWaveProcessor, getParallelProcessor };
export type { ChunkResult, ProcessResult, ProcessStats, CallbackFunction };

main().catch(err => {
  console.error(`${colors.red}Error:${colors.reset}`, err.message);
  process.exit(1);
});
